service/storage_service.cc

```diff
@@ -2658,20 +2658,207 @@ void storage_service::add_expire_time_if_found(inet_address endpoint, int64_t ex
+        while (!sstables.empty()) {
+            auto ops_uuid = utils::make_random_uuid();
+            auto sst_set = make_lw_shared<sstables::sstable_set>(sstables::make_partitioned_sstable_set(s,
+                    make_lw_shared<sstable_list>(sstable_list{}), false));
+            size_t batch_sst_nr = 16;
+            std::vector<sstring> sst_names;
+            std::vector<sstables::shared_sstable> sst_processed;
+            size_t estimated_partitions = 0;
+            while (batch_sst_nr-- && !sstables.empty()) {
+                auto sst = sstables.back();
+                estimated_partitions += sst->estimated_keys_for_range(full_token_range);
+                sst_names.push_back(sst->get_filename());
+                sst_set->insert(sst);
+                sst_processed.push_back(sst);
+                sstables.pop_back();
+            }
+
+            slogger.info("load_and_stream: started ops_uuid={}, process [{}-{}] out of {} sstables={}",
+                    ops_uuid, nr_sst_current, nr_sst_current + sst_processed.size(), nr_sst_total, sst_names);
 
-    slogger.info("Loading new SSTables for {}.{}...", ks_name, cf_name);
+            auto start_time = std::chrono::steady_clock::now();
+            lw_shared_ptr<dht::decorated_key> current_dk;
+            std::vector<gms::inet_address> current_targets;
+            std::unordered_map<gms::inet_address, send_meta_data> metas;
+            size_t num_partitions_processed = 0;
+            size_t num_bytes_read = 0;
+            nr_sst_current += sst_processed.size();
+            auto reader = table.make_streaming_reader(s, full_partition_range, sst_set);
 
-    return distributed_loader::process_upload_dir(_db, _sys_dist_ks, _view_update_generator, ks_name, cf_name).finally([this, ks_name, cf_name] {
-        slogger.info("Done loading new SSTables for {}.{}", ks_name, cf_name);
+            try {
+                netw::messaging_service& ms = _messaging.local();
+                for (;;) {
+                    auto mf = reader(db::no_timeout).get0();
+                    if (mf) {
+                        bool is_partition_start = mf->is_partition_start();
+                        if (is_partition_start) {
+                            ++num_partitions_processed;
+                            auto& start = mf->as_partition_start();
+                            current_dk = make_lw_shared<dht::decorated_key>(start.key());
+
+                            current_targets = rs.get_natural_endpoints(current_dk->token());
+                            if (primary_replica_only && current_targets.size() > 1) {
+                                current_targets.resize(1);
+                            }
+                            slogger.trace("load_and_stream: ops_uuid={}, current_dk={}, current_targets={}", ops_uuid,
+                                    current_dk->token(), current_targets);
+                            for (auto& node : current_targets) {
```

Comment:
What happens if we stream to ourself?

Reply:
Currently, it goes to the rpc like remote node. We can optimize the local write to avoid the rpc. I plan to do the optimization later.

Reply:
ok.

Reply:
(it's not bad, just want to be sure it works)

Reply:
It is bad if rpc compression is turned on. It is a waste to compress and uncompress locally.  I agree it is not urgent.